excellent/visitor.go

```diff
@@ -62,55 +65,63 @@ func (v *Visitor) VisitStringLiteral(ctx *gen.StringLiteralContext) interface{}
 	}
 
 	// replace "" with "
-	return strings.Replace(unquoted, "\"\"", "\"", -1)
+	unquoted = strings.Replace(unquoted, "\"\"", "\"", -1)
+
+	return types.NewXString(unquoted)
 }
 
 // VisitFunctionCall deals with function calls like TITLE(foo.bar)
 func (v *Visitor) VisitFunctionCall(ctx *gen.FunctionCallContext) interface{} {
 	functionName := strings.ToLower(ctx.Fnname().GetText())
 
-	var function XFunction
+	var function functions.XFunction
 	var found bool
 
-	function, found = XFUNCTIONS[functionName]
+	function, found = functions.XFUNCTIONS[functionName]
 	if !found {
-		return fmt.Errorf("no function with name '%s'", functionName)
+		return types.NewXErrorf("no function with name '%s'", functionName)
 	}
 
-	var params []interface{}
+	var params []types.XValue
 	if ctx.Parameters() != nil {
-		params = v.Visit(ctx.Parameters()).([]interface{})
+		params, _ = v.Visit(ctx.Parameters()).([]types.XValue)
 	}
 	val := function(v.env, params...)
 	return val
 }
 
 // VisitTrue deals with the "true" literal
 func (v *Visitor) VisitTrue(ctx *gen.TrueContext) interface{} {
-	return true
+	return types.XBoolTrue
 }
 
 // VisitFalse deals with the "false" literal
 func (v *Visitor) VisitFalse(ctx *gen.FalseContext) interface{} {
-	return false
+	return types.XBoolFalse
 }
 
 // VisitArrayLookup deals with lookups such as foo[5]
 func (v *Visitor) VisitArrayLookup(ctx *gen.ArrayLookupContext) interface{} {
-	context := v.Visit(ctx.Atom())
-	lookup, err := types.ToString(v.env, v.Visit(ctx.Expression()))
-	if err != nil {
-		return err
+	context := toXValue(v.Visit(ctx.Atom()))
```

Comment:
@nicpottier would be nice to replace these runtime checks with compile time checks